### PR TITLE
SymPy-powered distribution factory for NGBoost

### PR DESCRIPTION
**Pull request by guyko81**
## SymPy-powered distribution factory for NGBoost

A while back I proposed using SymPy during development to speed up adding new distributions. I showed two working examples - a simple one (Normal, where everything has closed form) and a harder one (Gamma, where the Fisher Information needs polygamma and string-based lambdify). That was a proof of concept with two demo scripts. 

Now I developed a full sympy-based factory, that generates everything for us, without the need of hard math.

### The factory

The core is `make_distribution()` in `ngboost/distns/sympy_utils.py`. You define a distribution with SymPy symbols and the factory derives everything else - gradients, Fisher Information, fitting, sampling, no manual calculus is needed. The resulting class is ready to pass directly to `NGBRegressor(Dist=...)`.

What it handles:
- Symbolic differentiation with automatic chain rule for log-transformed parameters
- Three-tier Fisher Information: closed-form Hessian, then SymPy expectation, then Monte Carlo fallback
- Numerically stable evaluation (overflow/NaN protection, logsumexp path for mixtures)
- Both regression and classification (`class_prob_exprs` for categorical outputs)
- Extra non-optimized parameters (like `n` in BetaBinomial)
- Scipy delegation for quantiles, CDF, sampling when available

For comparison, adding Beta by hand used to mean 100+ lines of digamma and polygamma expressions. Now the entire `beta.py` is 27 lines, basically just some imports and a single `make_distribution()` call. The factory detects that the Hessian doesn't depend on `y` and uses the closed-form Fisher Information directly.

### New distributions

I also added 5 new distributions based on this new factory, all importable from `ngboost.distns`:

| Distribution | Type | Use case |
|---|---|---|
| `Beta` | Regression | Bounded (0,1) outcomes |
| `LogitNormal` | Regression | Bounded (0,1) with logistic-normal model |
| `BetaBernoulli` | Classification | Binary with calibrated uncertainty |
| `BetaBinomial` | Regression | Overdispersed counts, known number of trials |
| `BetaBinomialEstN` | Regression | Overdispersed counts, unknown number of trials |

### Testing

To make sure I didn't make mistakes I added tests too:

- `test_sympy_existing_distns.py` generates SymPy versions of the existing Normal, Gamma, and Poisson and checks that `score()`, `d_score()`, and `metric()` match the hand-written implementations numerically. They match - which means the factory could in principle replace all existing score classes.
- All new distributions pass finite-difference gradient checks.
- Distributions with analytical Fisher Information (Beta, BetaBernoulli) pass Monte Carlo comparison tests.

### Example notebooks

I added 15 example notebooks for better understanding. Originally I only wanted to cover the new distributions, but decided to show all distributions for helping better understanding of the strength of NGBoost. Every distribution now has a self-contained runnable example - synthetic data, training, predictions, intervals, visualization. 

I added two additional notebooks, that demonstrate building custom distributions from scratch and working with mixture models. Please look at `sympy_normal_demo.ipynb` for the inner workings of the factory.

### Why did I do this

I always felt that the main bottleneck for contributing distributions to NGBoost is the manual math - getting derivatives right, handling chain rules for log-transformed parameters, computing or approximating Fisher Information. Now anyone who can express their distribution symbolically can contribute a working implementation or just use the tool directly without pull requests.

### CODE REVIEW

**Review by alejandroschuler — APPROVED** [review]

File: ngboost/distns/sympy_utils.py [hunk 1]
```diff
@@ -0,0 +1,832 @@
...
+        # already-lambdified score function.
+        _score_fn_for_fit = score_cls._score_fn  # pylint: disable=protected-access
+
+        def _fit_mle(Y):
+            def nll(internal):
+                natural = []
+                for v, (_, is_log) in zip(internal, param_info):
+                    natural.append(np.exp(np.clip(v, -150, _EXP_CLIP)) if is_log else v)
+                return np.mean(_score_fn_for_fit(Y, *natural))
+
```
Comment (ryan-wolbeck) [inline ngboost/distns/sympy_utils.py]: _fit_mle calls _score_fn_for_fit(Y, *natural) but never passes extra_params. Any distribution that uses extra_params in its score will crash at fit time when no fit_fn/scipy_dist_cls is provided. Please either thread extra_params (with defaults/kwargs) into _fit_mle, or raise a clear error if extra_params are present without a custom fit_fn or scipy_dist_cls.

```
    else:
        if _extra_param_info:
            raise ValueError(
                "extra_params are not supported by default _fit_mle; "
                "provide fit_fn or scipy_dist_cls/scipy_kwarg_map."
            )
        # Numerical MLE fallback: minimize mean NLL using the
        # already-lambdified score function.
        _score_fn_for_fit = score_cls._score_fn  # pylint: disable=protected-access
        ...
```

File: ngboost/distns/sympy_utils.py [hunk 2]
```diff
@@ -0,0 +1,832 @@
...
+        _fit = staticmethod(_fit_mle)
+
+    # ---- 5. sample ----
+    if sample_fn is not None:
+        _sample = sample_fn
+    elif scipy_dist_cls is not None:
+
+        def _sample(self, m):
+            return np.array([self.dist.rvs() for _ in range(m)])
+
```
Comment (ryan-wolbeck) [inline ngboost/distns/sympy_utils.py]: This should run faster here using vectorized sampling

```
def _sample(self, m):
    return np.asarray(self.dist.rvs(size=m))
```

File: ngboost/distns/sympy_utils.py [hunk 3]
```diff
@@ -0,0 +1,832 @@
...
+        def _class_probs(self):
+            param_arrays = [
+                np.asarray(getattr(self, pname), dtype=float) for pname, _ in param_info
+            ]
+            extra_arrays = [getattr(self, n) for n in _extra_names]
+            cols = [
+                np.asarray(fn(*param_arrays, *extra_arrays), dtype=float)
+                for fn in _class_prob_fns
+            ]
+            return np.column_stack(cols)
```
Comment (ryan-wolbeck) [inline ngboost/distns/sympy_utils.py]: minor: class_probs() just stacks the raw expressions. If numerics drift you can end up with probs outside [0,1] or rows not summing to 1. Might be worth clipping + normalizing here to keep downstream sampling/calibration sane.

```

def _class_probs(self):
    param_arrays = [
        np.asarray(getattr(self, pname), dtype=float) for pname, _ in param_info
    ]
    extra_arrays = [getattr(self, n) for n in _extra_names]
    cols = [
        np.asarray(fn(*param_arrays, *extra_arrays), dtype=float)
        for fn in _class_prob_fns
    ]
    probs = np.column_stack(cols)
    probs = np.clip(probs, 0.0, 1.0)
    row_sums = probs.sum(axis=1, keepdims=True)
    row_sums = np.where(row_sums == 0, 1.0, row_sums)
    return probs / row_sums

```

**Review by ryan-wolbeck — CHANGES_REQUESTED** [review]

File: ngboost/distns/sympy_utils.py [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+            return obj
+
+    else:
+        _getitem = None
+
+    # ---- 7. __getattr__ for scipy delegation ----
+    def _getattr(self, attr_name):
+        if "dist" in self.__dict__ and attr_name in dir(self.__dict__["dist"]):
+            return getattr(self.__dict__["dist"], attr_name)
+        return None
```
[the remark this answers is not in the document]
Comment (guyko81) [reply ngboost/distns/sympy_utils.py]: thanks, I tested it and it failed indeed. Fixing.

File: ngboost/distns/sympy_utils.py (hunk 1 above)
Comment (guyko81) [reply ngboost/distns/sympy_utils.py]: thanks, fixing

File: ngboost/distns/sympy_utils.py (hunk 2 above)
Comment (guyko81) [reply ngboost/distns/sympy_utils.py]: sample(): 0.1985s | rvs(size=m): 0.0006s | slowdown: 314.5x

well, it's convincing, thanks for this one too!
